drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -1066,6 +1068,38 @@ def _is_answer_correct(self):
         """
         return self._answer_correctness() == self.SOLUTION_CORRECT
 
+    def calculate_due_date(self):
+        """
+        Return the date submissions should be closed from.
+
+        Calculates and return the due date after which the submissions
+        will no longer be accepted. Returns None if both due & graceperiod
+        are missing which happens for pure DragAndDropXblock as they are
+        not its attributes. They are present for DragAndDropXblockWithMixins
+        """
+        if hasattr(self, 'due') and hasattr(self, 'graceperiod'):
+            due_date = self.due            # pylint: disable=no-member
+            if self.graceperiod is not None and due_date:          # pylint: disable=no-member
+                return due_date + self.graceperiod                 # pylint: disable=no-member
+            else:
+                return due_date
+        else:
+            return None
```

Comment:
It seems wrong that this sort of logic must be duplicated in every XBlock. What do you think about instead expanding and moving this `calculate_due_date()` method to `InheritanceMixin` (check `self_paced`, `due`, and `graceperiod`, and produce either a final due date or `None` if not applicable)?

Then this PR would be reduced to a few lines:

```
def student_view_data(self, context=None):
    ...
    "submit_deadline": self.calculate_due_date(),
    ...
```

and

```
    var problemNotDue = function () {
        if(configuration.submit_deadline === null){
            return true;
        }
        var now = new Date();
        var due_date = new Date(configuration.submit_deadline);
        return (now.getTime() < due_date.getTime());
    };
```

Reply:
I will have to consult other members of my team for this decision. Even though it makes sense to have this common function in the InheritanceMixin, we need to be sure about the consequences.